jre/javatests/com/google/j2cl/jre/java8/util/stream/CollectorsTest.java

```diff
@@ -672,4 +675,52 @@ void assertSingleItemCollectedAs(
     assertTrue("expected= " + expected + ", actual=" + actual, predicate.test(expected, actual));
   }
 
+  public void testFlatMapping() {
+    // since applyItems tests the same inputs multiple times, we need fresh stream instances as they can't be reused
+    Collector<Collection<String>, ?, List<String>> flatMapping = flatMapping(Collection::stream,
+        toList());
+    applyItems(Arrays.asList("a", "b"), flatMapping, Collections.singletonList("a"),
+        Collections.singletonList("b"));
+    applyItems(Arrays.asList("c", "d"), flatMapping, Collections.emptyList(), Arrays.asList("c", "d"));
+        
+    Collector<Collection<String>, ?, List<String>> flatMappingToNull = flatMapping(items -> {
+        if (items.size() % 2 == 0) {
+          // Return null instead of empty to test behavior around null
+          return null;
+        }
+        return items.stream();
+    }, toList());
+    applyItems(Arrays.asList("a"), flatMappingToNull, Arrays.asList("a"), Arrays.asList("b", "c"));
+
+    Collector<Collection<String>, ?, List<String>> flatMappingToNullWithEmptyList = flatMapping(items -> {
+        if (items.size() == 0) {
+            // Return null instead of empty to test behavior around null
+            return null;
+        }
+        return items.stream();
+    }, toList());
+    applyItems(Arrays.asList("a", "b", "c"), flatMappingToNullWithEmptyList, Collections.emptyList(), Arrays.asList("a", "b", "c"));
+
+
+    Stream<Integer> stream = Arrays.asList(1, 2, 3).stream();
+    stream.collect(flatMapping(x -> Stream.of(x, x), toList()));
+
+    IllegalStateException thrown = null;
+    try {
+      stream.anyMatch(x -> x > 0);
```

Comment:
You are checking original stream is closed; not the stream flatMapped to. You can return a custom stream from map function and ensure "close" is called on it.

Reply:
@gkdn do you mean something like this:

```
        List<Stream<Integer>> streams = new ArrayList<>();
        Stream.of(1, 2, 3).collect(flatMapping(x -> {
            Stream<Integer> current = Stream.of(x, x);
            streams.add(current);
            return current;
        }, toList()));

        assertEquals(3, streams.size());

        for (Stream<Integer> current : streams) {
            IllegalStateException thrown = null;
            try {
                current.anyMatch(x -> x > 0);
            } catch (IllegalStateException e) {
                thrown = e;
            }
            assertNotNull(thrown);
            assertEquals("stream has already been operated upon or closed", thrown.getMessage());
        }
```

Reply:
I would do something like :

```
 int[] calledCount = {0};
 Stream<Object> mapped = Stream.of("x").onClose(x -> calledCount[0]++);
 Stream.of(1).collect(flatMapping(x -> mapped, toList()));
 assertEquals(1, calledCount[0]);
```

Reply:
@gkdn oh, sorry, I didn’t understand what you wanted me to fix, let me update the issue. 
Thank you